### DISCUSSION TITLE
Allow clients to modify or override generated Bazel targets for Swift packages

### DISCUSSION THREAD

**Discussion by cgrindel** [cgrindel/rules_swift_package_manager#852] [Ideas]
## Problem

Occasionally, a client tries to use a Swift package and the generated Bazel targets are not quite right or broken. Today, the client submits a bug report and someone updates `rules_swift_package_manager` to fix the issue. The problem is that the cycle for the fix can be long.

> [!NOTE]
> We still want the bug report, if a client runs into a problem. The goal is to provide a way for the client to work around the issue.

## Proposal

Create a mechanism by which a client can specify a target and apply fixes in much the same way as [buildozer](https://github.com/bazelbuild/buildtools/blob/master/buildozer/README.md). In fact, we should use `buildozer` wherever we can.

The mechanism needs a selector syntax and a modification syntax. `buildozer` already has these. 🥳 For the purposes of this discussion, I believe that the modification syntax should be whatever `buildozer` supports. Unfortunately, we may need to augment the selector syntax as we need to support third-party repository labels or something that approximates them.

### Usage

```starlark
swift_deps = use_extension(
    "@rules_swift_package_manager//:extensions.bzl",
    "swift_deps",
)

# A single buildozer command
swift_deps.buildozer(
    swift_package = "sdwebimage",
    cmd = "set testonly True",
    labels = ["//foo:bar", "//cake:chocolate"],
)

# Use a file with builozer commands as supported by -f
swift_deps.buildozer(
    swift_package = "sdwebimage",
    file = "//third-party/:sdwebimage.bldzr",
)
```

### Use patches

This would work but is fairly brittle. Any changes to how `rules_swift_package_manager` generates the underlying code could break client's patches.

## Conclusion

Let me know what you think. Do you like the idea? How would you want it to work?

**Comment by luispadron** [comment]
This is nice and would help with some edge cases we currently work around with by just not letting the plugin generate the build file. 

Would this replace the option that exists today to allow patching a package? It may not be needed if we have this

**Reply by cgrindel** [reply]
> Would this replace the option that exists today to allow patching a package?

I think that the patch system seems useful for source patching as @jpsim pointed out.

**Comment by jpsim** [comment]
TIL about the [existing patching mechanism](https://github.com/cgrindel/rules_swift_package_manager/blob/main/docs/patch_swift_package.md), thanks for mentioning it!

I think the proposed `swift_deps.buildozer` extension would be really nice, and would be a good place to make build-level changes, like flipping `testonly` or silencing some build warnings in 3rd parties via `copts`.

However, the `swift_pkg_patches.yaml` feature would still be best for making source-level changes, like if there's a bug or behavioral tweak in the third party you want to patch.
